kotlinx-coroutines-core/common/test/sync/MutexTest.kt

```diff
@@ -148,4 +148,20 @@ class MutexTest : TestBase() {
         assertFailsWith<IllegalStateException> { mutex.lock(owner) }
         assertFailsWith<IllegalStateException> { select { mutex.onLock(owner) {} } }
     }
+
+    @Test
+    fun testWithLockJsMiscompilation() = runTest {
+        // This is a reproducer for KT-58685
+        // On Kotlin/JS IR, the compiler miscompiles calls to 'unlock' in an inlined finally
+        // This is visible on the withLock function
+        // Until the compiler bug is fixed, this test case checks that we do not suffer from it
+        val mutex = Mutex()
+        assertFailsWith<IndexOutOfBoundsException> {
+            try {
+                mutex.withLock { null } ?: throw IndexOutOfBoundsException() // should throw…
+            } catch (e: Exception) {
+                throw e // …but instead fails here
+            }
+        }
+    }
```

Comment:
I think we need a complementary test, like `testJsMiscompilationIsStillHappening`, that would check that with an alternative, naturally-written definition of `withLock`, anomalous behavior *does* happen. This way, when the bad behavior stops, the test will fail and we'll know that the workaround is no longer needed. This will also ensure that `testWithLockJsMiscompilation` doesn't just work accidentally, we'll have a clear distinction between "this version clearly works" and "that version clearly doesn't." Does this make sense?

Reply:
If we add this, won't it means that whenever it is fixed on the compiler side, it will break CI until someone rolls back this PR?

I'll try to do this using the reproducer from the issue.

Reply:
Actually, thinking back on this, I'm not sure it's so easy. This function is inline. The bug happens at the call site.

Therefore, even if the KotlinX.Coroutines team uses a compiler that doesn't have the issue anymore, a user of the library could still be using an older compiler that has the issue, right?

Reply:
Yeah, the point about CI makes sense, given that the issue is already fixed on the compiler's side, so we don't need a notification.

Regarding the old compiler + new library version: we don't provide guarantees as to what happens when using a library compiled with a newer compiler than your code uses. Generally, it's error-prone.

So, let's not add the tests. I wrote one just to make sure the tests that you added *do* correctly show that there's no problem, and sure enough, your tests do invoke the miscompilation if we keep the old form of `withLock`/`withPermit`. Nice work!

```kotlin
class SemaphoreTest: TestBase() {
    @Test
    fun testWithPermitJsMiscompilation() = runTest {
        // This is a reproducer for KT-58685
        // On Kotlin/JS IR, the compiler miscompiles calls to 'release' in an inlined finally
        // This is visible on the withPermit function
        // Until the compiler bug is fixed, this test case checks that we do suffer from it
        val semaphore = Semaphore(1)
        assertFailsWith<IllegalStateException> {
            try {
                semaphore.withPermitMiscompiled { null } ?: throw IndexOutOfBoundsException() // should throw…
            } catch (e: Exception) {
                throw e // …but instead fails here
            }
        }
    }
}

suspend inline fun <T> Semaphore.withPermitMiscompiled(action: () -> T): T {
    acquire()
    try {
        return action()
    } finally {
        release()
    }
}
```